lib/matplotlib/ticker.py

```diff
@@ -823,49 +874,68 @@ def label_minor(self, labelOnlyBase):
         Switch minor tick labeling on or off.
 
-        ``labelOnlyBase=True`` to turn off minor ticks.
+        Parameters
+        ----------
+        labelOnlyBase : bool
+            If True, label ticks only at integer powers of base.
+
         """
         self.labelOnlyBase = labelOnlyBase
 
-    def set_locs(self, locs):
+    def set_locs(self, locs=None):
+        """
+        Use axis view limits to control which ticks are labeled.
+
+        The ``locs`` parameter is ignored in the present algorithm.
+
+        """
+        if np.isinf(self.minor_thresholds[0]):
+            self._sublabels = None
+            return
+
         b = self._base
 
         vmin, vmax = self.axis.get_view_interval()
         self.d = abs(vmax - vmin)
 
-        if not hasattr(self.axis, 'get_transform'):
-            # This might be a colorbar dummy axis, do not attempt to get
-            # transform
-            numdec = 10
-        elif hasattr(self.axis.get_transform(), 'linthresh'):
-            t = self.axis.get_transform()
-            linthresh = t.linthresh
+        # Handle symlog case:
+        linthresh = self._linthresh
+        if linthresh is None:
+            try:
+                linthresh = self.axis.get_transform().linthresh
+            except AttributeError:
+                pass
+
+        if linthresh is not None:  # symlog
             # Only compute the number of decades in the logarithmic part of the
             # axis
             numdec = 0
             if vmin < -linthresh:
-                numdec += math.log(-vmin / linthresh) / math.log(b)
-
-            if vmax > linthresh and vmin < linthresh:
-                numdec += math.log(vmax / linthresh) / math.log(b)
-            elif vmin >= linthresh:
-                numdec += math.log(vmax / vmin) / math.log(b)
+                rhs = min(vmax, -linthresh)
+                numdec += math.log(vmin / rhs) / math.log(b)
+            if vmax > linthresh:
+                lhs = max(vmin, linthresh)
+                numdec += math.log(vmax / lhs) / math.log(b)
         else:
             vmin = math.log(vmin) / math.log(b)
             vmax = math.log(vmax) / math.log(b)
             numdec = abs(vmax - vmin)
 
-        if numdec > 1:
+        if numdec > self.minor_thresholds[0]:
             # Label only bases
-            self.sublabel = set((1,))
+            self._sublabels = set((1,))
+        elif numdec > self.minor_thresholds[1]:
+            # Add labels between bases at log-spaced coefficients;
+            # include base powers in case the locations include
+            # "major" and "minor" points, as in colorbar.
+            c = np.logspace(0, 1, b//2 + 1, base=b)
+            self._sublabels = set(np.round(c))
         else:
-            # Add labels between bases at log-spaced coefficients
-            c = np.logspace(0, 1, (4 - int(numdec)) + 1, base=b)
-            self.sublabel = set(np.round(c))
+            self._sublabels = set(np.linspace(1, b, b))
```

Comment:
See #7583 for a fix.  Contrary to my comment above, I decided it is OK to allow sublabels when b is not an integer, though I expect this will be extremely rare in practice.